### PR TITLE
Compatibility to protobuf-net v3

### LINKED ISSUE #66: Incompatibility with protobuf-net v3
Hi,

We have recently updated to protobuf-net v3.0.29 and now get the following exception

```
"System.MissingMethodException: Method not found: 'Void ProtoBuf.ProtoWriter..ctor(System.IO.Stream, ProtoBuf.Meta.TypeModel, ProtoBuf.SerializationContext)'.

Server stack trace: 
   at ProtoBuf.Data.ProtoDataWriter.Serialize(Stream stream, IDataReader reader, ProtoDataWriterOptions options)
   at ProtoBuf.Data.DataSerializerEngine.Serialize(Stream stream, DataSet dataSet, ProtoDataWriterOptions options)
   at Brady.Etrm.Base.Infrastructure.Remoting.DataSetBinaryRemotingFormatSerializationSurrogate.GetObjectData(Object obj, SerializationInfo info, StreamingContext context) in C:\Development\etrm\Source\ECS\Libraries\Viz.Base\Infrastructure\Remoting\DataSetBinaryRemotingFormatSerializationSurrogate.cs:line 28
   at Brady.Etrm.Base.Infrastructure.Remoting.RootObjectProtobufSerializationWrapper.GetObjectData(SerializationInfo info, StreamingContext context) in C:\Development\etrm\Source\ECS\Libraries\Viz.Base\Infrastructure\Remoting\RootObjectProtobufSerializationWrapper.cs:li
ne 34
   at System.Runtime.Serialization.Formatters.Binary.WriteObjectInfo.InitSerialize(Object obj, ISurrogateSelector surrogateSelector, StreamingContext context, SerObjectInfoInit serObjectInfoInit, IFormatterConverter converter, ObjectWriter objectWriter, SerializationBinder binder)
   at System.Runtime.Serialization.Formatters.Binary.WriteObjectInfo.Serialize(Object obj, ISurrogateSelector surrogateSelector, StreamingContext context, SerObjectInfoInit serObjectInfoInit, IFormatterConverter converter, ObjectWriter objectWriter, SerializationBinder binder)
   at System.Runtime.Serialization.Formatters.Binary.ObjectWriter.Serialize(Object graph, Header[] inHeaders, __BinaryWriter serWriter, Boolean fCheck)
   at System.Runtime.Serialization.Formatters.Binary.BinaryFormatter.Serialize(Stream serializationStream, Object graph, Header[] headers, Boolean fCheck)
   at System.Runtime.Serialization.Formatters.Binary.BinaryFormatter.Serialize(Stream serializationStream, Object graph, Header[] headers
)
   at Brady.Etrm.Infrastructure.Remoting.BinaryWithProtobufSerializationHelper.SerializeBinaryMessage(IMessage msg, Stream outputStream) in C:\Development\etrm\Source\ECS\Libraries\Viz.Base\Infrastructure\Remoting\BinaryWithProtobufSerializationHelper.cs:line 63
   at Brady.Etrm.Infrastructure.Remoting.BradyServerFormatterSink.SerializeResponse(IServerResponseChannelSinkStack sinkStack, IMessage msg, ITransportHeaders& headers, Stream& stream) in C:\Development\etrm\Source\ECS\Libraries\Viz.Base\Infrastructure\Remoting\BradyServerFormatterSink.cs:line 170
   at Brady.Etrm.Infrastructure.Remoting.BradyServerFormatterSink.ProcessMessage(IServerChannelSinkStack sinkStack, IMessage requestMsg, ITransportHeaders requestHeaders, Stream requestStream, IMessage& responseMsg, ITransportHeaders& responseHeaders, Stream& responseStream) in C:\Development\etrm\Source\ECS\Libraries\Viz.Base\Infrastructure\Remoting\BradyServerFormatterSink.cs:line 95

Exception rethrown at [0]:
   at System.Runtime.Remoting.Proxies.RealProxy.HandleReturnMessage(IMessage reqMsg, IMessage retMsg)
   at System.Runtime.Remoting.Proxies.RealProxy.PrivateInvoke(MessageData& msgData, Int32 type)
   at Viz.Middleware.Extractor.VizPeer.IExtractorDataSetService.GetUserPermissionList(String user, String database)
   at Viz.EamSuite.DealManager.Database.DatabaseAccessPoint.GetUserPermissions() in C:\Development\etrm\Source\ECS\EFM\Viz.Efm.Presentation.Model\ViewModels\DealManager\DealWindow\Database\DatabaseAccessPoint.cs:line 854
   at Viz.EamSuite.DealManager.ViewPortfolioMonitor.RefreshPermissionList() in C:\Development\etrm\Source\ECS\EFM\Viz.Efm.Presentation.Windows\Views\DealManager\ContractManagement\Views\ViewPortfolioMonitor.cs:line 7840"
```

I see from [protobuf v3 docs](https://github.com/protobuf-net/protobuf-net/blob/main/docs/3_0.md) that:

> ProtoReader/ProtoWriter must now be instantiated via ProtoReader.Create, not new (recent v2 releases have been advising this change for some time); most users will not be using these APIs.

Are there any plans to support v3 or can a new version be published to cap at v2 so that our nuget refs can detect and cap the protobuf-net at latest v2?

Thanks

### PR DESCRIPTION
**Pull request by thomasjoscht**
Hey,

I've implemented compatiblity to protobuf-net v3. Main points:
- NuGet Upgrade to protobuf-net v3
- Changed Targets to at least net461(prerequisite for v3)
- TreatWarningsAsErrors=false because of many Obsolete annotations
- Fixed Problems with v3 like using static Create() methods and call writer.Close() before Dispose

All tests have passed successful (only one has to be skipped ColumnsReaderTests.TheReadColumnsMethod.ShouldThrowExceptionOnColumnsInvalidFieldHeader). 

Could you please have a look? Should fix #66

Regards Thomas

### CODE REVIEW

**Comment by UFuz** [comment]
Please one of the responsible can merge and close this PR? Highly appreciated!

**Comment by dotarj** [comment]
Hey @thomasjoscht, thanks for the PR. As you probably have noticed, I’m currently not really active on GitHub due to other activities (that will hopefully change). I got a message from @rdingwall about this PR. He suggested to make you contributor for this project so you can complete this PR. That’s fine for me. I did put some comments in the PR though. Maybe you can take a look at these comments?

I guess the NuGet key for AppVeyor is expired. I will update the key so that the CI pipeline can publish the artefacts to NuGet.

**Review by UFuz — APPROVED** [review]

**Review by dotarj — CHANGES_REQUESTED** [review]

**Comment by dotarj** [comment]
@UFuz Ah, I did not actually submit the comments. You should see them now. I could take a look myself, but I’m currently away from home (holiday).

**Comment by thomasjoscht** [comment]
Hey @dotarj, sorry for my late response. I needed v3 compatibility very quickly for one of my projects and thought to share it with you as PR. Currently my time is limited and I'm unable to do a full review. Sorry about that. Perhaps if we need more changes in future I'll dive deeper in your code.

+1 for merge and sharing via NuGet

File: src/protobuf-net-data/protobuf-net-data.csproj [hunk 1]
```diff
@@ -6,8 +6,8 @@
     <DebugType Condition="$(codecov) != 'true'">embedded</DebugType>
     <GenerateDocumentationFile>true</GenerateDocumentationFile>
     <IncludeSymbols>false</IncludeSymbols>
-    <TargetFrameworks>net20;net30;net35;net40;net45;netstandard2.0</TargetFrameworks>
-    <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
+    <TargetFrameworks>net461;netstandard2.0</TargetFrameworks>
+    <TreatWarningsAsErrors>false</TreatWarningsAsErrors>
```
[the remark this answers is not in the document]
Comment (thomasjoscht) [reply src/protobuf-net-data/protobuf-net-data.csproj:10]: This was necessary because of many warnings about obsolete methods. Some small typo but also some larger problems with static methods of ProtoWriter and ProtoReader like Create, StartSubItem and EndSubItem.

File: test/protobuf-net-data.Tests/ColumnsReaderTests.TheReadColumnsMethod.cs [hunk 2]
```diff
@@ -12,19 +12,21 @@ public partial class ColumnsReaderTests
     {
         public class TheReadColumnsMethod : ColumnsReaderTests
         {
-            [Fact]
+            [Fact(Skip = "Upgrade to protobuf-net v3 will raise InvalidOperationException : The writer is in an incomplete state during close")]
```
[the remark this answers is not in the document]
Comment (thomasjoscht) [reply test/protobuf-net-data.Tests/ColumnsReaderTests.TheReadColumnsMethod.cs:15]: This could lead to a problem. In my case everything is working just fine but an uncertainty is possible. I haven't enough time for testing and fixing this Tests. Sorry ...

**Comment by dxdjgl** [comment]
+1 for merge and sharing via NuGet

**Comment by CorrieMeyer** [comment]
Hi @dotarj , will there be updated nuget for this change? I also need, 3.0 compatibility with protobuf-net ?

**Comment by dotarj** [comment]
Hi @CorrieMeyer, with the current state of this pull request it can’t be completed. I’ve taken a look and it seems that the threat warnings as errors was mainly disabled due to a lot of obsolete warnings (and some other warnings). The API of protobuf-net has changed. With backwards compatibility for now, but also a lot of obsolete messages. I think it would be better to suppress the obsolete messages for now and enable the warnings as errors again. Using the new API would require a lot more work and can be done in the future. The disabled unit test looks like an edge case but I will need to do some more testing on that (or if someone else would like to do that that would be great).

@thomasjoscht, can you [enable the option “Allow edits from maintainers”](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork) on your pull request so I can push changes to your branch?

**Comment by thomasjoscht** [comment]
Hi @dotarj, the option "Allow edits by maintainers" is enabled for this pull request. Do you need anything more?

**Comment by dotarj** [comment]
Hi @thomasjoscht, thanks for settings the option. It worked, I just pushed some changes to the branch.

File: test/protobuf-net-data.Tests/protobuf-net-data.Tests.csproj [hunk 3]
```diff
@@ -3,8 +3,8 @@
   <PropertyGroup>
     <DebugType Condition="$(codecov) == 'true'">full</DebugType>
     <IsPackable>false</IsPackable>
-    <TargetFrameworks>net40;net471;netcoreapp2.0</TargetFrameworks>
-    <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
+    <TargetFrameworks>net461;net471;netcoreapp2.0</TargetFrameworks>
+    <TreatWarningsAsErrors>false</TreatWarningsAsErrors>
```
[the remark this answers is not in the document]
Comment (dotarj) [reply test/protobuf-net-data.Tests/protobuf-net-data.Tests.csproj]: Warnings were fixed and threat warnings as errors re-enabled.

File: src/protobuf-net-data/protobuf-net-data.csproj (hunk 1 above)
[the remark this answers is not in the document]
Comment (dotarj) [reply src/protobuf-net-data/protobuf-net-data.csproj]: Warnings were fixed and threat warnings as errors re-enabled.

**Comment by dotarj** [comment]
It has taken a while but there is a new version available on NuGet: https://github.com/dotarj/protobuf-net-data/releases/tag/4.0.0.
